MoveInfo rearrangement and flag optimisation

cfmnephrite opened the pull request:
## Description
Basically, a reordering of the BattleMove struct into a more sensible, more efficient structure.
- Grouped BP/accuracy and type/category together (finally)
- Removed unnecessary flags
    - Made some into EFFECT flags
    - Rewrote code to take modified flags into account
    - Notably, removed Sheer Force flag and used additionalEffects field to determine if a move should be boosted or not
- removed numAdditionalEffects and used some bit trickery to fit a counter for up to 15 additional effects inside the `additionalEffects` field
- Rearranged all moves_info.h data to be to be in struct order

## **Discord contact info**
The Cancer Fairy/thechurchofcage

AlexOn1ine commented on test/battle/move_effect/upper_hand.c:
```diff
@@ -99,12 +99,12 @@ SINGLE_BATTLE_TEST("Upper Hand fails if the target moves first")
 }
 
-SINGLE_BATTLE_TEST("Upper Hand is boosted by Sheer Force")
+SINGLE_BATTLE_TEST("Upper Hand is boosted by Sheer Force") // why does this move need a special test?
```
It's because it can't be added together with all the other sheer force tests.

Should have asked in the dev channel. Now you have to waste time removing the comment. 😅

AlexOn1ine replied:
Also conficts

cfmnephrite commented:
Conflicts resolved and `moves_info.h` information all sorted in struct order (since CONFIG_BY_GEN isn't likely to be resolved by release day). Ready for review :)

AlexOn1ine commented:
@AsparagusEduardo I think this should make it into 1.8, not 1.9 since this is the last change for the move refactor.

cfmnephrite commented:
Forgive the very late update, had another look and all of a sudden, BP/accuracy/type/category looked wrong to me compared to type/category/BP/accuracy :P It also matches their order on places like Showdown and Bulbapedia (except PP, but oh, well).

AsparagusEduardo commented:
> @AsparagusEduardo I think this should make it into 1.8, not 1.9 since this is the last change for the move refactor.

Alright, I'll take a look before release.

AlexOn1ine commented:
> > @AsparagusEduardo I think this should make it into 1.8, not 1.9 since this is the last change for the move refactor.
> 
> Alright, I'll take a look before release.

If you want I can review it. I was holding back because I thought you might want to take a look.

AsparagusEduardo commented on test/battle/ability/sheer_force.c:
```diff
@@ -1,14 +1,20 @@
 #include "global.h"
 #include "test/battle.h"
 
+
+ASSUMPTIONS
+{
+    ASSUME(MoveIsAffectedBySheerForce(MOVE_ELECTRO_SHOT) == TRUE);
+}
+
```
Why an assumption specifically for Electro Shot?

cfmnephrite replied:
Yup, Electro Shot uses a "hack" to be Sheer Force boosted so it's just a check for that.

AsparagusEduardo commented on src/data/moves_info.h:
```diff
@@ -122,15 +122,15 @@ const struct MoveInfo gMovesInfo[MOVES_COUNT_DYNAMAX] =
         .priority = 0,
-        .category = DAMAGE_CATEGORY_PHYSICAL,
-        .metronomeBanned = TRUE,
         .mirrorMoveBanned = TRUE,
```
Ok, I need to ask this.

*Why?*

Why change the stablished order? I get grouping banned flags together, but this will create so many conflicts across the board that I don't get why you chose to do this.

cfmnephrite replied:
To put everything into a more sensible order and make better use of space. You approved this when I asked about it ages ago?

cfmnephrite replied:
![Screenshot_20240213-081302_Discord (1)](https://github.com/rh-hideout/pokeemerald-expansion/assets/22523466/f07db5ec-18c8-47c7-b158-e96ff8614f84)

Tbf, I think you were more interested in the flag rearrangement, but all of the above was in the proposal, for which you said this ^

cfmnephrite replied:
Besides, we've cause conflicts before by renaming `split` (and then also the defines - twice), adding contest data and names, and adding `additionalEffects`. This would probably add more conflicts than any of those but it would be worth it to establish, IMO, now that all move data fields have been added to the struct, a final, clean, sensible order.

AsparagusEduardo replied:
> Tbf, I think you were more interested in the flag rearrangement, but all of the above was in the proposal, for which you said this ^

From what I remember, you only showed the struct rearrangement, not that you were going to change the field order in the data as well.

AsparagusEduardo replied:
> Besides, we've cause conflicts before by renaming `split` (and then also the defines - twice), adding contest data and names, and adding `additionalEffects`. This would probably add more conflicts than any of those but it would be worth it to establish, IMO, now that all move data fields have been added to the struct, a final, clean, sensible order.

The current order is fine. Vanilla fields on top, new fields at the bottom.

cfmnephrite replied:
Oh! I thought that having moves info and the struct be in synch was preferred. Has that not been the case?

cfmnephrite replied:
In fact, I seem to recall with the species form refactor how you didn't like that the order of the fields had been lost (and you were satisfied when I then send that I would rearrange them again before finishing).

cfmnephrite replied:
> The current order is fine to me. Vanilla fields on top, new fields at the bottom.

They aren't really, tbh. The zMove field is way up there, and the flags are in kind of a random order as well (not that I'm willing to touch those but still).

AsparagusEduardo replied:
> In fact, I seem to recall with the species form refactor how you didn't like that the order of the fields had been lost (and you were satisfied when I then said that I would rearrange them again before finishing).

I don't recall this much. In the end, the struct was arranged to minimize space used but then the fields were organized such that Dex data and graphics data were all next to each other, despite not that being how it's setup in the struct.

cfmnephrite replied:
@AsparagusEduardo This is what you said back then and it's why I spent my time reordering moves_info to match the struct:
![Screenshot_20240213-090257_Edge (1)](https://github.com/rh-hideout/pokeemerald-expansion/assets/22523466/e4955bf8-b520-466a-9cc1-f860c08f877e)

AsparagusEduardo replied:
Because this was an automated thing you did that moved fields that were previously in species macros to the top of the array. All mon that didn't have macros kept their previous order, so it was inconsistent.

AsparagusEduardo replied:
In fact, the macros I made were curated to keep the field order that I had established for all non-macro Pokémon, so base stats on top and so on.

AsparagusEduardo replied:
I need to gather my thoughts. I've been outside of the house the whole day and I can't focus.

cfmnephrite replied:
Reverted the change since we can do it later anyway

AsparagusEduardo commented on include/pokemon.h:
```diff
@@ -451,44 +451,40 @@ struct MoveInfo
-
+    u16 type:5;
+    u16 category:2;
+    u16 power:9; // up to 511
```
Is there a move with more than 255 power?

cfmnephrite replied:
Not in vanilla, but I don't see why we shouldn't allow for that possibility in romhacks. Besides, it's not like we need up to 255 accuracy.

AsparagusEduardo replied:
Then we don't need to move the extra bit.

cfmnephrite replied:
I can't without reordering things again

AsparagusEduardo commented on include/pokemon.h:
```diff
@@ -451,44 +451,40 @@ struct MoveInfo
+    u32 strikeCount:4; // Max 15 hits. Defaults to 1 if not set. May apply its effect on each hit.
+    u32 criticalHitStage:2;
+    u32 alwaysCriticalHit:1;
+    // 14 bits left to complete this word - continues into flags
```
Add this padding as a field.

cfmnephrite replied:
It's not padding though. Like it says, it just continues into flags. It's just a note for anyone who would like to add an extra field in that position.

AsparagusEduardo replied:
You could add a space to delimit where the word end instead (or a comment)

AsparagusEduardo commented on src/battle_ai_main.c:
```diff
@@ -2507,7 +2507,6 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
 
                 if (instructedMove == MOVE_NONE
                   || gMovesInfo[instructedMove].instructBanned
-                  || gMovesInfo[instructedMove].twoTurnMove
```
Shouldn't this still check if it's a 2-turn move?

AlexOn1ine replied:
I think the idea is for the AI to charge and fire the move in the same turn if instructed so it would make sense to remove it.

ZnogyroP replied:
Two-turn moves are fully banned from Instruct, even if it could be executed it one turn.

AlexOn1ine replied:
oh then it even makes more sense to remove it.

AsparagusEduardo replied:
> oh then it even makes more sense to remove it.

What? It makes more sense to *not* remove it in this case then, replaced by the effect twoTurnMove flag

AlexOn1ine replied:
I mean hopefully those moves have `instructBanned` on them already

AsparagusEduardo commented on include/battle.h:
```diff
@@ -54,11 +54,15 @@ struct __attribute__((packed, aligned(2))) BattleMoveEffect
     u16 semiInvulnerableEffect:1;
-    u16 flags:11; // coming soon...
+    u16 usesProtectCounter:1;
+    u16 forcePressure:1;
```
I feel `forcePressure` should be set for moves instead of effects.

AsparagusEduardo replied:
Which means that it could be interpreted either way, and when that happens, I rather keep it for moves to allow more customization.

AlexOn1ine commented:
CF, your changes look great overall but since Edu isn't sure on the reorder in `moves_info.h` could you revert that part? It would be nice to get this merged before 1.8.

AlexOn1ine commented:
> CF, your changes look great overall but since Edu isn't sure on the reorder in `moves_info.h` could you revert that part? It would be nice to get this merged before 1.8.

+ conflicts :eyes:

cfmnephrite commented:
@AlexOn1ine Done, just waiting for CI

AlexOn1ine commented:
Rename suggestions. 
```
RETURN_MOVE_HAS_MOVE_EFFECT_WITH -> RETURN_HAS_MOVE_ADDITIONAL_EFFECT
MoveHasMoveEffect -> MoveHasAdditionalEffect
MoveHasMoveEffectWithChance -> MoveHasAdditionalEffectWithChance 
MoveHasMoveEffectSelf -> MoveHasAdditionalEffectSelf
MoveHasMoveEffectSelfArg -> MoveHasAdditionalEffectSelfArg
MoveHasChargeTurnMoveEffect -> MoveHasChargeTurnAdditionalEffect
```

cfmnephrite commented:
> MoveHasMoveEffect

Done

AlexOn1ine reviewed (approved):
@AsparagusEduardo From my side it can be merged. The review was mostly done on discord.

AsparagusEduardo commented on test/battle/move.c:
```diff
@@ -23,13 +23,12 @@ SINGLE_BATTLE_TEST("Accuracy controls the proportion of misses")
+    PARAMETRIZE { move = MOVE_THUNDER_SHOCK; chance = 10; }
+    PARAMETRIZE { move = MOVE_DISCHARGE; chance = 30; }
+    PARAMETRIZE { move = MOVE_NUZZLE; chance = 100; }
+    ASSUME(MoveHasAdditionalEffectWithChance(move, MOVE_EFFECT_PARALYSIS, chance) == TRUE);
```
Hmm, shouldn't this test technically still pass if the secondary effect chances are changed by the user? From the name of the test, I assume we want to make sure that the chance modifier does change the chances, so I don't think hardcoding the chances in the assume here is a good idea.
```suggestion
    ASSUME(MoveHasAdditionalEffect(move, MOVE_EFFECT_PARALYSIS) == TRUE);
```

AsparagusEduardo commented on include/pokemon.h:
```diff
@@ -529,15 +526,23 @@ struct MoveInfo
+#define ADDITIONAL_EFFECTS(...) ((min(ARRAY_COUNT(EFFECTS_ARR( __VA_ARGS__ )), 15)) << 28) + (uintptr_t)(EFFECTS_ARR( __VA_ARGS__ ))
+
+// Retrieve a move's additional effects and the count thereof
+#define GET_ADDITIONAL_EFFECTS_AND_COUNT(move, _count, _effects) u32 _count = GET_ADDITIONAL_EFFECTS_COUNT(move); const struct AdditionalEffect *_effects = GET_ADDITIONAL_EFFECTS(move)
+#define GET_ADDITIONAL_EFFECTS(move) (void *)(gMovesInfo[move].additionalEffects & 0x8FFFFFF)
+#define GET_ADDITIONAL_EFFECTS_COUNT(move) (gMovesInfo[move].additionalEffects >> 28)
```
I recommend moving `GET_ADDITIONAL_EFFECTS_AND_COUNT` to the bottom.
```suggestion
#define GET_ADDITIONAL_EFFECTS(move) (void *)(gMovesInfo[move].additionalEffects & 0x8FFFFFF)
#define GET_ADDITIONAL_EFFECTS_COUNT(move) (gMovesInfo[move].additionalEffects >> 28)
#define GET_ADDITIONAL_EFFECTS_AND_COUNT(move, _count, _effects) u32 _count = GET_ADDITIONAL_EFFECTS_COUNT(move); const struct AdditionalEffect *_effects = GET_ADDITIONAL_EFFECTS(move)
```

AsparagusEduardo replied:
oh, and rename `GET_ADDITIONAL_EFFECTS_COUNT` to `GET_ADDITIONAL_EFFECT_COUNT`

AsparagusEduardo commented on src/battle_util.c:
```diff
@@ -11082,56 +11079,50 @@ bool32 IsGen6ExpShareEnabled(void)
+
+/* Quick way of checking if a move has move effects with match a few
+comma-separated conditions. Each condition has to check a field of AdditionalEffect. */
+#define RETURN_HAS_MOVE_ADDITIONAL_EFFECT(condition1, ...)           \
```
```suggestion
#define RETURN_HAS_MOVE_ADDITIONAL_EFFECT(condition1, ...)          \
```

cfmnephrite replied:
What am I supposed to be seeing here? 🤔 
![Uploading Screenshot_20240226-205859_Edge (1).jpg…]()

cfmnephrite replied:
What  I supposed to be changing here?
![Screenshot_20240226-205859_Edge (1)](https://github.com/rh-hideout/pokeemerald-expansion/assets/22523466/07fce8ba-b71b-44d9-a133-606b2ebd1f1f)

AsparagusEduardo replied:
Alignment
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/2904965/8ff29ba9-164b-4ea6-90bd-05de7611c3b4)

AsparagusEduardo reviewed (changes requested):
Besides some small adjustements, looks good to me :)

cfmnephrite commented:
Suggestions done

AsparagusEduardo commented:
PR is not passing on the latest upcoming

cfmnephrite commented:
It looks like TIMEOUT isn't caught by KNOWN_FAILING, for whatever reason. Just fixed the test and removed KNOWN_FAILING.

mrgriffin commented:
> It looks like TIMEOUT isn't caught by KNOWN_FAILING, for whatever reason. Just fixed the test and removed KNOWN_FAILING.

Glad it's fixed! The reason why `TIMEOUT` isn't caught is because t's not known whether it's a pass or failure, because we don't know that if we ran the test for long enough it wouldn't eventually pass (i.e. it's a variant of the halting problem). Or from an alternative POV, it's not a failure: it's an error.

cfmnephrite commented:
> > It looks like TIMEOUT isn't caught by KNOWN_FAILING, for whatever reason. Just fixed the test and removed KNOWN_FAILING.
> 
> Glad it's fixed! The reason why `TIMEOUT` isn't caught is because it's not known whether it's a pass or failure, because we don't know that if we ran the test for long enough it wouldn't eventually pass (i.e. it's a variant of the halting problem). Or from an alternative POV, it's not a failure: it's an error.

If it's uncertain whether or not it has passed, it should absolutely be treated as a failure IMO. But I'm not willing to address it technically so I don't care, lol.

In any case, fixed and ready for review :)
